File: GameLibrary/src/cc/lib/checkerboard/DragonChess.java
package cc.lib.checkerboard;

import cc.lib.game.Utils;
import cc.lib.utils.Table;

import static cc.lib.checkerboard.Game.FAR;
import static cc.lib.checkerboard.Game.NEAR;
import static cc.lib.checkerboard.PieceType.BISHOP;
import static cc.lib.checkerboard.PieceType.BLOCKED;
import static cc.lib.checkerboard.PieceType.DRAGON_IDLE_L;
import static cc.lib.checkerboard.PieceType.DRAGON_IDLE_R;
import static cc.lib.checkerboard.PieceType.EMPTY;
import static cc.lib.checkerboard.PieceType.FLAG_DRAGON;
import static cc.lib.checkerboard.PieceType.KNIGHT_L;
import static cc.lib.checkerboard.PieceType.KNIGHT_R;
import static cc.lib.checkerboard.PieceType.PAWN_IDLE;

public class DragonChess extends Chess {

    @Override
    void init(Game game) {
        whiteSide = Utils.flipCoin() ? FAR : NEAR;
        // this is to enforce the 'queen on her own color square' rule
        PieceType left = PieceType.QUEEN;
        PieceType right = PieceType.UNCHECKED_KING_IDLE;
        if (whiteSide == FAR) {
            right = PieceType.QUEEN;
            left = PieceType.UNCHECKED_KING_IDLE;
        }

        game.init(10, 8+6);
        game.initRank(0, FAR, BLOCKED, BLOCKED, BLOCKED, DRAGON_IDLE_R, KNIGHT_R, BISHOP, left, right, BISHOP, KNIGHT_L, DRAGON_IDLE_L, BLOCKED, BLOCKED, BLOCKED);
        game.initRank(1, FAR, BLOCKED, BLOCKED, BLOCKED, PAWN_IDLE, PAWN_IDLE, PAWN_IDLE, PAWN_IDLE, PAWN_IDLE, PAWN_IDLE, PAWN_IDLE, PAWN_IDLE, BLOCKED, BLOCKED, BLOCKED);
        game.initRank(2, -1  , BLOCKED, BLOCKED, BLOCKED, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY,EMPTY, EMPTY, EMPTY, BLOCKED, BLOCKED, BLOCKED);
        game.initRank(3, -1  , EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY);
        game.initRank(4, -1  , EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY);
        game.initRank(5, -1  , EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY);
        game.initRank(6, -1  , EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY);
        game.initRank(7, -1  , BLOCKED, BLOCKED, BLOCKED, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, BLOCKED, BLOCKED, BLOCKED);
        game.initRank(8, NEAR, BLOCKED, BLOCKED, BLOCKED, PAWN_IDLE, PAWN_IDLE, PAWN_IDLE, PAWN_IDLE, PAWN_IDLE, PAWN_IDLE, PAWN_IDLE, PAWN_IDLE, BLOCKED, BLOCKED, BLOCKED);
        game.initRank(9, NEAR, BLOCKED, BLOCKED, BLOCKED, DRAGON_IDLE_R, KNIGHT_R, BISHOP, left, right, BISHOP, KNIGHT_L, DRAGON_IDLE_L, BLOCKED, BLOCKED, BLOCKED);

        game.setTurn(whiteSide);

    }

    @Override
    protected int[] getRookCastleCols(Game game) {
        return new int [] { 3, 3+7 };
    }

    @Override
    protected boolean isSquareAttacked(Game game, int rank, int col, int attacker) {
        if (super.isSquareAttacked(game, rank, col, attacker))
            return true;

        int kn=computeKDN(rank, col, DELTAS_N, DELTAS_S, DELTAS_E, DELTAS_W, DELTAS_NE, DELTAS_NW, DELTAS_SE, DELTAS_SW);
        for (int k=0; k<kn; k++) {
            int [][] kd = kdn[k];
            int [] dr = kd[0];
            int [] dc = kd[1];
            int num = Math.min(3, dr.length);
            for (int i = 0; i < num; i++) {
                int rr = rank + dr[i];
                int cc = col + dc[i];
                if (testPiece(game, rr, cc, attacker, FLAG_DRAGON))
                    return true;
                if (game.getPiece(rr, cc).getType() != EMPTY)
                    break;
            }
        }

        return false;
    }

    @Override
    Table getInstructions() {
        return new Table("Chess Variation. Bigger board, non square\nand using Dragons instead of rooks.\nDragons move like a queen but with a max of\n3 squares.");
    }
}

File: Checkerboard/src/cc/lib/checkerboard/DragonChess.kt
package cc.lib.checkerboard

import cc.lib.game.Utils
import cc.lib.utils.Table

class DragonChess : Chess() {
	override fun init(game: Game): Array<Array<Piece>> {
		whiteSide = if (Utils.flipCoin()) Game.FAR else Game.NEAR
		// this is to enforce the 'queen on her own color square' rule
		var left = PieceType.QUEEN
		var right = PieceType.UNCHECKED_KING_IDLE
		if (whiteSide == Game.FAR) {
			right = PieceType.QUEEN
			left = PieceType.UNCHECKED_KING_IDLE
		}
		game.turn = whiteSide
		return initFromPieceTypes(arrayOf(
			Game.FAR to arrayOf(PieceType.BLOCKED, PieceType.BLOCKED, PieceType.BLOCKED, PieceType.DRAGON_IDLE_R, PieceType.KNIGHT_R, PieceType.BISHOP, left, right, PieceType.BISHOP, PieceType.KNIGHT_L, PieceType.DRAGON_IDLE_L, PieceType.BLOCKED, PieceType.BLOCKED, PieceType.BLOCKED),
			Game.FAR to arrayOf(PieceType.BLOCKED, PieceType.BLOCKED, PieceType.BLOCKED, PieceType.PAWN_IDLE, PieceType.PAWN_IDLE, PieceType.PAWN_IDLE, PieceType.PAWN_IDLE, PieceType.PAWN_IDLE, PieceType.PAWN_IDLE, PieceType.PAWN_IDLE, PieceType.PAWN_IDLE, PieceType.BLOCKED, PieceType.BLOCKED, PieceType.BLOCKED),
			Game.NOP to arrayOf(PieceType.BLOCKED, PieceType.BLOCKED, PieceType.BLOCKED, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.BLOCKED, PieceType.BLOCKED, PieceType.BLOCKED),
			Game.NOP to arrayOf(PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY),
			Game.NOP to arrayOf(PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY),
			Game.NOP to arrayOf(PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY),
			Game.NOP to arrayOf(PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY),
			Game.NOP to arrayOf(PieceType.BLOCKED, PieceType.BLOCKED, PieceType.BLOCKED, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.EMPTY, PieceType.BLOCKED, PieceType.BLOCKED, PieceType.BLOCKED),
			Game.NEAR to arrayOf(PieceType.BLOCKED, PieceType.BLOCKED, PieceType.BLOCKED, PieceType.PAWN_IDLE, PieceType.PAWN_IDLE, PieceType.PAWN_IDLE, PieceType.PAWN_IDLE, PieceType.PAWN_IDLE, PieceType.PAWN_IDLE, PieceType.PAWN_IDLE, PieceType.PAWN_IDLE, PieceType.BLOCKED, PieceType.BLOCKED, PieceType.BLOCKED),
			Game.NEAR to arrayOf(PieceType.BLOCKED, PieceType.BLOCKED, PieceType.BLOCKED, PieceType.DRAGON_IDLE_R, PieceType.KNIGHT_R, PieceType.BISHOP, left, right, PieceType.BISHOP, PieceType.KNIGHT_L, PieceType.DRAGON_IDLE_L, PieceType.BLOCKED, PieceType.BLOCKED, PieceType.BLOCKED),
		))
	}

	override fun getRookCastleCols(game: Game): IntArray {
		return intArrayOf(3, 3 + 7)
	}

	override fun isSquareAttacked(game: Game, rank: Int, col: Int, attacker: Int): Boolean {
		if (super.isSquareAttacked(game, rank, col, attacker)) return true
		val kn = computeKDN(rank, col, DELTAS_N, DELTAS_S, DELTAS_E, DELTAS_W, DELTAS_NE, DELTAS_NW, DELTAS_SE, DELTAS_SW)
		for (k in 0 until kn) {
			val kd = kdn[k]
			val dr = kd[0]
			val dc = kd[1]
			val num = Math.min(3, dr.size)
			for (i in 0 until num) {
				val rr = rank + dr[i]
				val cc = col + dc[i]
				if (testPiece(game, rr, cc, attacker, PieceType.FLAG_DRAGON)) return true
				if (game.getPiece(rr, cc).getType() !== PieceType.EMPTY) break
			}
		}
		return false
	}

	override val instructions: Table
		get() = Table().addRow("Chess Variation. Bigger board, non square and using Dragons instead of rooks. Dragons move like a queen but with a max of 3 squares.")
}